src/mscorlib/shared/System/Convert.cs

```diff
@@ -2656,44 +2656,134 @@ public static bool TryFromBase64String(string s, Span<byte> bytes, out int bytes
-                bytesWritten = 0;
-                return true;
-            }
+            bytesWritten = 0;
 
-            // We need to get rid of any trailing white spaces.
-            // Otherwise we would be rejecting input such as "abc= ":
-            while (chars.Length > 0)
+            while (chars.Length != 0)
             {
-                char lastChar = chars[chars.Length - 1];
-                if (lastChar != ' ' && lastChar != '\n' && lastChar != '\r' && lastChar != '\t')
+                // Attempt to decode a segment that doesn't contain whitespace.
+                bool complete = Base64.TryDecodeFromUtf16(chars, bytes, out int consumedInThisIteration, out int bytesWrittenInThisIteration);
+                bytesWritten += bytesWrittenInThisIteration;
+                if (complete)
+                    return true;
+
+                chars = chars.Slice(consumedInThisIteration);
+                bytes = bytes.Slice(bytesWrittenInThisIteration);
+
+                Debug.Assert(chars.Length != 0); // If TryDecodeFromUtf16() consumed the entire buffer, it could not have returned false.
+                if (chars[0].IsSpace())
                 {
-                    break;
-                }
-                chars = chars.Slice(0, chars.Length - 1);
-            }
+                    // If we got here, the very first character not consumed was a whitespace. We can skip past any consecutive whitespace, then continue decoding.
 
-            fixed (char* charsPtr = &MemoryMarshal.GetReference(chars))
-            {
-                int resultLength = FromBase64_ComputeResultLength(charsPtr, chars.Length);
-                Debug.Assert(resultLength >= 0);
-                if (resultLength > bytes.Length)
+                    int indexOfFirstNonSpace = 1;
+                    for (; ; )
+                    {
+                        if (indexOfFirstNonSpace == chars.Length)
+                            break;
+                        if (!chars[indexOfFirstNonSpace].IsSpace())
+                            break;
+                        indexOfFirstNonSpace++;
+                    }
+
+                    chars = chars.Slice(indexOfFirstNonSpace);
+
+                    if ((bytesWrittenInThisIteration % 3) != 0 && chars.Length != 0)
+                    {
+                        // If we got here, the last successfully decoded block encountered an end-marker, yet we have trailing non-whitespace characters.
+                        // That is not allowed.
+                        bytesWritten = default;
+                        return false;
+                    }
+
+                    // We now loop again to decode the next run of non-space characters. 
+                }
+                else
                 {
-                    bytesWritten = 0;
-                    return false;
+                    unsafe
+                    {
+                        Debug.Assert(chars.Length != 0 && !chars[0].IsSpace());
```

Comment:
Is there a way to refactor the code here to avoid such deep nesting (maybe pull whatever is in the else case into a method)? As it stands, it is difficult to reason about what's happening exactly just by inspection.

Reply:
We have a number of loop variables that can be updated (chars, bytes, bytesWritten, etc.) I'm not sure it'd be much simpler to understand. For sure, this method is at the limit of my complexity tolerance - that's why the tempBuffer copy code is in its own method.